### PR TITLE
Make task queues configurable

### PR DESCRIPTION
**Pull request by Diogenesoftoronto**
refs #621

### CODE REVIEW

File: cmd/enduro-a3m-worker/main.go [hunk 1]
```diff
@@ -56,6 +57,13 @@ func main() {
 		fmt.Printf("Failed to read configuration: %v\n", err)
 		os.Exit(1)
 	}
+	// Set Defaults for the a3m and global task queues
+	if reflect.ValueOf(cfg.A3m.TaskQueue).IsZero() {
+		cfg.A3m.TaskQueue = temporal.A3mWorkerTaskQueue
+	}
+	if reflect.ValueOf(cfg.Temporal.TaskQueue).IsZero() {
+		cfg.A3m.TaskQueue = temporal.GlobalTaskQueue
+	}
```
Comment (sevein) [inline cmd/enduro-a3m-worker/main.go]: I haven't had time to review this PR before going on vacation, I just have one comment here: using the reflect package to check the zero value is probably overkill and slower. The idiomatic way is: `if x == "" { ... }`.

Comment (sevein) [reply cmd/enduro-a3m-worker/main.go]: If the config now contains the task queues I don't think we need to maintain `temporal.A3mWorkerTaskQueue` nor `temporal.GloblaTaskQueue`. You should probably use Viper to set defaults or have the application fail because these configuration attributes are mandatory and not allowed to be undefined.

Comment (Diogenesoftoronto) [reply cmd/enduro-a3m-worker/main.go]: defaults are now set by viper. I did keep the values there just instead of raw strings because it makes it easy to track things down via lsp

File: internal/package_/workflow.go [hunk 2]
```diff
@@ -81,9 +83,10 @@ func InitProcessingWorkflow(ctx context.Context, tc temporalsdk_client.Client, r
 }
 
 type MoveWorkflowRequest struct {
-	ID         uint
-	AIPID      string
-	LocationID uuid.UUID
+	ID              uint
+	AIPID           string
+	LocationID      uuid.UUID
+	GlobalTaskQueue string
```
Comment (sevein) [inline internal/package_/workflow.go]: This should read `TaskQueue`.

File: internal/package_/package_.go [hunk 3]
```diff
@@ -39,18 +39,20 @@ type packageImpl struct {
 	evsvc          event.EventService
 	tokenVerifier  auth.TokenVerifier
 	ticketProvider *auth.TicketProvider
+	TaskQueue      string
 }
 
 var _ Service = (*packageImpl)(nil)
 
-func NewService(logger logr.Logger, db *sql.DB, tc temporalsdk_client.Client, evsvc event.EventService, tokenVerifier auth.TokenVerifier, ticketProvider *auth.TicketProvider) *packageImpl {
+func NewService(logger logr.Logger, db *sql.DB, tc temporalsdk_client.Client, evsvc event.EventService, tokenVerifier auth.TokenVerifier, ticketProvider *auth.TicketProvider, taskqueue string) *packageImpl {
```
Comment (sevein) [inline internal/package_/package_.go]: It should read `taskQueue` instead of `taskqueue`. I would also suggest to list it right after the function parameter that receives the Temporal client:

    func NewService(logger logr.Logger, db *sql.DB, tc temporalsdk_client.Client, taskQueue string, evsvc event.EventService, tokenVerifier auth.TokenVerifier, ticketProvider *auth.TicketProvider) *packageImpl {

File: internal/config/config.go [hunk 4]
```diff
@@ -60,6 +60,8 @@ func Read(config *Configuration, configFile string) (found bool, configFileUsed
 	v.AddConfigPath("/etc")
 	v.SetConfigName("enduro")
 	v.SetDefault("api.processing", a3m.ProcessingDefault)
+	v.SetDefault("a3m.taskqueue", temporal.A3mWorkerTaskQueue)
+	v.SetDefault("temporal.taskqueue", temporal.A3mWorkerTaskQueue)
```
Comment (sevein) [inline internal/config/config.go]: I was wondering if we should do something like the following:

```go
v.SetDefault("a3m.taskqueue", a3m.DefaultTaskQueue)
v.SetDefault("temporal.taskqueue", temporal.DefaultTaskQueue)
```

(pls notice the change in the variable name: DefaultXXX)

That way we push to the a3m package the concern of naming its default task queue.

I'm also realizing that the `temporal` package has now an oudated code comment:

```
const (
	// There are task queues used by our workflow and activity workers. It may
	// be convenient to make these configurable in the future .
	GlobalTaskQueue    = "global"
	A3mWorkerTaskQueue = "a3m"
)
```

File: internal/package_/goa.go [hunk 5]
```diff
@@ -251,9 +251,10 @@ func (w *goaWrapper) Move(ctx context.Context, payload *goapackage.MovePayload)
 	}
 
 	_, err = InitMoveWorkflow(ctx, w.tc, &MoveWorkflowRequest{
-		ID:         payload.ID,
-		AIPID:      *goapkg.AipID,
-		LocationID: payload.LocationID,
+		ID:              payload.ID,
+		AIPID:           *goapkg.AipID,
+		LocationID:      payload.LocationID,
+		GlobalTaskQueue: w.TaskQueue,
```
Comment (sevein) [inline internal/package_/goa.go]: `TaskQueue` instead of `GlobalTaskQueue`.

File: internal/workflow/processing.go [hunk 6]
```diff
@@ -74,6 +73,9 @@ type TransferInfo struct {
 	//
 	// It is populated by createPreservationActionLocalActivity .
 	PreservationActionID uint
+
+	GlobalTaskQueue string
+	A3mTaskQueue    string
```
Comment (sevein) [inline internal/workflow/processing.go:81]: These two also deserve a comment.

File: internal/package_/workflow.go [hunk 7]
```diff
@@ -60,6 +58,10 @@ type ProcessingWorkflowRequest struct {
 
 	// Location identifier for storing auto approved AIPs.
 	DefaultPermanentLocationID *uuid.UUID
+
+	// The global task queue used for starting the workflow
```
Comment (sevein) [inline internal/package_/workflow.go]: Missing period at the end of the sentence. Maybe something like the following?

```suggestion
	// Task queues used for starting new workflows.
```

File: internal/storage/service.go [hunk 8]
```diff
@@ -161,7 +161,7 @@ func (s *serviceImpl) Update(ctx context.Context, payload *goastorage.UpdatePayl
 
 	signal := UploadDoneSignal{}
 	workflowID := fmt.Sprintf("%s-%s", StorageUploadWorkflowName, aipID)
-	err = s.tc.SignalWorkflow(ctx, workflowID, "", UploadDoneSignalName, signal)
+	err = s.tc.SignalWorkflow(ctx, workflowID, s.config.TaskQueue, UploadDoneSignalName, signal)
```
Comment (sevein) [inline internal/storage/service.go]: I'm confused about this change, the third parameter is a `runID` that we're meant to leave empty in this specific context. We don't need the name of the task queue when signaling a workflow, the workflow ID will be enough.

File: internal/storage/service.go [hunk 9]
```diff
@@ -218,7 +219,7 @@ func (s *serviceImpl) MoveStatus(ctx context.Context, payload *goastorage.MoveSt
 		return nil, err
 	}
 
-	resp, err := s.tc.DescribeWorkflowExecution(ctx, fmt.Sprintf("%s-%s", StorageMoveWorkflowName, p.AipID), "")
+	resp, err := s.tc.DescribeWorkflowExecution(ctx, fmt.Sprintf("%s-%s", StorageMoveWorkflowName, p.AipID), s.config.TaskQueue)
```
Comment (sevein) [inline internal/storage/service.go]: Same here, the `runID` must be left empty.

Task queues and run identifiers are different concepts in Temporal.

File: internal/storage/service_test.go [hunk 10]
```diff
@@ -191,18 +192,18 @@ func TestServiceSubmit(t *testing.T) {
 		attrs := &setUpAttrs{}
 		svc := setUpService(t, attrs)
 		ctx := context.Background()
-
+		// attrs.config.TaskQueue = "global"
```
Comment (sevein) [inline internal/storage/service_test.go]: Should this be removed?

File: internal/storage/service_test.go [hunk 11]
```diff
@@ -164,11 +165,11 @@ func TestServiceSubmit(t *testing.T) {
 				mock.AnythingOfType("*context.timerCtx"),
 				temporalsdk_client.StartWorkflowOptions{
 					ID:                    "storage-upload-workflow-" + AIPID.String(),
-					TaskQueue:             "global",
+					TaskQueue:             "",
```
Comment (sevein) [inline internal/storage/service_test.go]: I don't understand your choice here, why would you want to pass an empty string? This is happening more than once in this file.

File: internal/storage/workflow.go [hunk 12]
```diff
@@ -21,12 +19,14 @@ const (
 )
 
 type StorageUploadWorkflowRequest struct {
-	AIPID uuid.UUID
+	AIPID           uuid.UUID
+	GlobalTaskQueue string
```
Comment (sevein) [inline internal/storage/workflow.go]: Use `TaskQueue`.

File: internal/storage/workflow.go [hunk 13]
```diff
@@ -21,12 +19,14 @@ const (
...
 }
 
 type StorageMoveWorkflowRequest struct {
-	AIPID      uuid.UUID
-	LocationID uuid.UUID
+	AIPID           uuid.UUID
+	LocationID      uuid.UUID
+	GlobalTaskQueue string
```
Comment (sevein) [inline internal/storage/workflow.go]: Use `TaskQueue`.

File: internal/workflow/move.go [hunk 14]
```diff
@@ -9,14 +9,16 @@ import (
 )
 
 type MoveWorkflow struct {
-	logger logr.Logger
-	pkgsvc package_.Service
+	logger          logr.Logger
+	pkgsvc          package_.Service
+	globalTaskQueue string
```
Comment (sevein) [inline internal/workflow/move.go]: `taskQueue`

**Review by sevein — COMMENTED** [review]
I've suggested a number of minor modifications. I realize there are quite a few and I hope it's not overwhelming.

**Comment by Diogenesoftoronto** [comment]
Hey, this is ready to review again. I added all of your suggestions.

**Comment by sevein** [comment]
There is a failing test:

```
--- FAIL: TestServiceMoveStatus (0.01s)
    --- FAIL: TestServiceMoveStatus/Returns_move_done_if_workflow_completed (0.00s)
```

**Comment by Diogenesoftoronto** [comment]
It works!

File: go.mod [hunk 15]
```diff
@@ -137,14 +137,14 @@ require (
...
+	golang.org/x/mod v0.13.0 // indirect
 	golang.org/x/net v0.17.0 // indirect
 	golang.org/x/oauth2 v0.12.0 // indirect
-	golang.org/x/sync v0.3.0 // indirect
+	golang.org/x/sync v0.4.0 // indirect
 	golang.org/x/sys v0.13.0 // indirect
 	golang.org/x/text v0.13.0 // indirect
 	golang.org/x/time v0.3.0 // indirect
-	golang.org/x/tools v0.13.0 // indirect
+	golang.org/x/tools v0.14.0 // indirect
```
Comment (sevein) [inline go.mod]: It'd be nice if you don't include unrelated changes to go.mod in your commit if they're not strictly related.

File: internal/package_/package_.go [hunk 16]
```diff
@@ -39,18 +39,20 @@ type packageImpl struct {
...
 	return &packageImpl{
 		logger:         logger,
 		db:             sqlx.NewDb(db, "mysql"),
 		tc:             tc,
 		evsvc:          evsvc,
 		tokenVerifier:  tokenVerifier,
 		ticketProvider: ticketProvider,
+		taskQueue:      taskqueue,
```
Comment (sevein) [inline internal/package_/package_.go]: It should read `taskQueue` instead of `taskqueue`.

File: internal/workflow/activities/bundle.go [hunk 17]
```diff
@@ -133,7 +133,7 @@ func (a *BundleActivity) SingleFile(ctx context.Context, transferDir, key, tempF
 		return "", fmt.Errorf("error moving file (from %s to %s): %v", tempFile, path, err)
 	}
 
-	if err := os.Chmod(path, os.FileMode(0o755)); err != nil {
+	if err := os.Chmod(path, os.FileMode(0o755)); err != nil { //#nosec G301 -- Are group execute perms required?
```
Comment (sevein) [inline internal/workflow/activities/bundle.go]: This is still needed? Looks unrelated to your changes.

File: internal/workflow/move.go [hunk 18]
```diff
@@ -9,14 +9,16 @@ import (
...
 }
 
-func NewMoveWorkflow(logger logr.Logger, pkgsvc package_.Service) *MoveWorkflow {
+func NewMoveWorkflow(logger logr.Logger, pkgsvc package_.Service, taskqueue string) *MoveWorkflow {
 	return &MoveWorkflow{
-		logger: logger,
-		pkgsvc: pkgsvc,
+		logger:    logger,
+		pkgsvc:    pkgsvc,
+		taskQueue: taskqueue,
```
Comment (sevein) [inline internal/workflow/move.go]: It should read `taskQueue` instead of `taskqueue`.

File: main.go [hunk 19]
```diff
@@ -192,6 +192,7 @@ func main() {
 	// Set up the storage service.
 	var storagesvc storage.Service
 	{
+		cfg.Storage.TaskQueue = cfg.Temporal.TaskQueue
```
Comment (sevein) [inline main.go]: What would be the motivation of doing this?

**Review by sevein — COMMENTED** [review]
A few small suggestions/questions.

File: main.go (hunk 19 above)
Comment (Diogenesoftoronto) [reply main.go]: i have no idea why i did this it does not seem to cause any bugs when i remove it.

File: internal/workflow/activities/bundle.go (hunk 17 above)
Comment (Diogenesoftoronto) [reply internal/workflow/activities/bundle.go]: Tried to run gosec locally to see if this was really needed, could not get it to work at first. But I finally got it working now and this is not necessary anymore.

**Comment by sevein** [comment]
Just realized this is ready for another review, will do tomorrow! Sorry for the late response.

**Comment by Diogenesoftoronto** [comment]
> Just realized this is ready for another review, will do tomorrow! Sorry for the late response.

Perfectly fine. Should have sent you a DM.

**Review by sevein — APPROVED** [review]
GitHub won't allow me to comment on the specific line, but the `storage.Submit` method (L117) was not updated to pass the task queue to the upload workflow request. I had to make this change:

```diff
-       _, err = InitStorageUploadWorkflow(ctx, s.tc, &StorageUploadWorkflowRequest{AIPID: aipID})
+       _, err = InitStorageUploadWorkflow(ctx, s.tc, &StorageUploadWorkflowRequest{
+               AIPID:     aipID,
+               TaskQueue: s.config.TaskQueue,
+       })
```

Other than that, :+1:. I'll merge, thank you!

This link shows the changes I had to make but it's confusing because it also includes the rebase work I did in preparation for the merge: [compare:0f3b/60ee](https://github.com/artefactual-sdps/enduro/compare/0f3b50fbe6c07977deb8732405a44cee8faec0bd..60ee7f93667d57194cbe876f4f5ae9f6001f5eb2). The relevant bits are in `internal/storage/service.go` and `internal/storage/service_test.go` in case you're interested.
